control_software/src/paper_gpu/file_conversion.py

```diff
@@ -0,0 +1,349 @@
+        field, both 32-bit integers.
+
+    Raises
+    ------
+    ValueError
+        Raised if the data read in cannot be reshaped into the specified shape.
+    """
+    # read raw binary data
+    data = np.fromfile(filename, dtype=_hera_corr_dtype)
+    try:
+        data = data.reshape(data_shape)
+    except ValueError:
+        raise ValueError(
+            f"data cannot be reshaped; data read is {data.size} elements, "
+            "target size is {data_shape}"
+        )
+
+    # return to user
+    return data
+
+
+def get_antpos_info():
+    """
+    Fetch HERA antenna positions from hera_mc.
+
+    Parameters
+    ----------
+    None
+
+    Returns
+    -------
+    antpos_xyz : ndarray
+        An array of floats of size (350, 3) which contains the antenna
+        positions in XYZ (i.e., ECEF) coordinates.
+    ant_names : ndarray
+        An array of strings of size (350,) which contains the antenna names.
+    """
+    # read antenna positions from M&C
+    ants = geo_sysdef.read_antennas()
+    # convert from eastings/northings to ENU
+    # HERA is in Zone 34J; corresponds to latitude 10000000 in northings
+    latlon_p = ccrs.Geodetic()
+    utm_p = ccrs.UTM(34)
+    lat_corr = 10000000
+    antpos_xyz = np.empty((350, 3), dtype=np.float64)
+    ant_names = np.empty((350,), dtype="S5")
+    for ant, pos in ants.items():
+        # unpack antenna information
+        antnum = int(ant[2:])
+        if antnum > 350:
+            # skip bizarre HT701 entry
+            continue
+        easting = pos["E"]
+        northing = pos["N"]
+        elevation = pos["elevation"]
+        # transform coordinates to ECEF
+        lon, lat = latlon_p.transform_point(easting, northing - lat_corr, utm_p)
+        xyz = uvutils.XYZ_from_LatLonAlt(np.radians(lat), np.radians(lon), elevation)
+        antpos_xyz[antnum, :] = xyz
+        # also save antenna names
+        ant_names[antnum] = ant
+
+    return antpos_xyz, ant_names
+
+
+def make_uvh5_file(filename, metadata_file, data_file):
+    """
+    Make a UVH5 file from a metdata + raw binary data file.
+
+    This function creates a valid UVH5 file from the specified metadata and
+    binary data files. It adds the flags and nsample datasets, and applies
+    bitshuffle compression to the data.
+
+    Parameters
+    ----------
+    filename : str
+        The name of the output file to write.
+    metadata_file : str
+        The name of the metadata file written by the correlator.
+    data_file : str
+        The name of the data file written by the correlator.
+
+    Returns
+    -------
+    None
+    """
+    # get cminfo from redis
+    cminfo = redis_cm.read_cminfo_from_redis(return_as="dict")
+
+    # get antennas positions and names from mc
+    antpos_xyz, ant_names = get_antpos_info()
+    nants_telescope = len(ant_names)
+
+    # read in metadata
+    metadata = read_header_data(metadata_file)
```

Comment:
suggest read_header_data return dict that labels the metadata, rather than use position, which divides knowledge of header format between read_header and make_uvh5_file